src/ir/AST/AST.hs

```diff
@@ -586,6 +622,12 @@ data Expr = Skip {emeta :: Meta Expr}
                          args :: Arguments}
           | Optional {emeta :: Meta Expr,
                       optTag :: OptionalPathComponent}
+          | AdtExtractorPattern {emeta :: Meta Expr,
+                                 ty :: Type,
+                                 name :: Name,
+                                 arg :: Expr,
+                                 fieldNames :: [String],
+                                 adtClassDecl :: ClassDecl}
```

Comment:
Does this have to be a `ClassDecl`, or could it be a type that is looked up when needed?